### PR TITLE
Annotate ObjectPool with nullable

### LINKED ISSUE #5680: Update ASP.NET Core to use C# 8's nullable reference types
Updating ASP.NET Core to use C# 8's nullable reference types would:

1. Help ASP.NET Core libraries avoid null reference exceptions internally. It will help us find and prevent our bugs and increase our developer productivity
2. Provide guidance to developers who are using ASP.NET Core about which APIs can accept and return a null reference and which APIs can't. This would improve the developer experience of using ASP.NET Core

@rynowak @davidfowl @DamianEdwards @Eilon 

----------------------
## Projects in no particular order

* [x] Microsoft.AspNetCore.DataProtection.EntityFrameworkCore
* [x] Microsoft.AspNetCore.DataProtection.StackExchangeRedis
* [x] Microsoft.AspNetCore.TestHost
* [x] Microsoft.AspNetCore.Hosting.WindowsServices
* [ ] Microsoft.AspNetCore.ApiAuthorization.IdentityServer
* [x] Microsoft.AspNetCore.Identity.EntityFrameworkCore
* [ ] Microsoft.AspNetCore.Identity.Specification.Tests
* [x] Microsoft.AspNetCore.Identity.UI
* [x] Microsoft.AspNetCore.Authentication.Certificate
* [x] Microsoft.AspNetCore.Authentication.Facebook
* [x] Microsoft.AspNetCore.Authentication.Google
* [x] Microsoft.AspNetCore.Authentication.JwtBearer
* [x] Microsoft.AspNetCore.Authentication.MicrosoftAccount
* [x] Microsoft.AspNetCore.Authentication.Negotiate
* [x] Microsoft.AspNetCore.Authentication.OpenIdConnect
* [x] Microsoft.AspNetCore.Authentication.Twitter
* [x] Microsoft.AspNetCore.Authentication.WsFederation
* [ ] Microsoft.Extensions.Logging.AzureAppServices
* [x] Microsoft.AspNetCore.ConcurrencyLimiter
* [x] Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore
* [x] Microsoft.AspNetCore.HeaderPropagation
* [x] Microsoft.Extensions.Diagnostics.HealthChecks.EntityFrameworkCore
* [x] Microsoft.AspNetCore.MiddlewareAnalysis
* [x] Microsoft.AspNetCore.SpaServices.Extensions
* [x] Microsoft.AspNetCore.Mvc.NewtonsoftJson
* [x] Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation
* [x] Microsoft.AspNetCore.Mvc.Testing
* [x] Microsoft.AspNetCore.AzureAppServicesIntegration
* [x] Microsoft.AspNetCore.SignalR.Client.Core
* [x] Microsoft.AspNetCore.SignalR.Client
* [x] Microsoft.AspNetCore.Http.Connections.Client
* [x] Microsoft.AspNetCore.SignalR.Protocols.MessagePack
* [x] Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson
* [ ] Microsoft.AspNetCore.SignalR.Specification.Tests
* [x] Microsoft.AspNetCore.SignalR.StackExchangeRedis
* [x] Microsoft.Authentication.WebAssembly.Msal
* [x] Microsoft.JSInterop.WebAssembly
* [x] Microsoft.AspNetCore.Components.WebAssembly.Server
* [x] Microsoft.AspNetCore.Components.WebAssembly.Authentication
* [x] Microsoft.AspNetCore.Components.WebAssembly
* [x] Microsoft.AspNetCore
* [x] Microsoft.AspNetCore.DataProtection.Abstractions
* [x] Microsoft.AspNetCore.Cryptography.Internal
* [x] Microsoft.AspNetCore.Cryptography.KeyDerivation
* [x] Microsoft.AspNetCore.DataProtection
* [x] Microsoft.AspNetCore.DataProtection.Extensions
* [x] Microsoft.AspNetCore.Antiforgery
* [x] Microsoft.AspNetCore.Hosting.Abstractions
* [x] Microsoft.AspNetCore.Hosting
* [x] Microsoft.AspNetCore.Hosting.Server.Abstractions
* [x] Microsoft.AspNetCore.Authentication.Abstractions
* [x] Microsoft.AspNetCore.Authentication.Core
* [x] Microsoft.Net.Http.Headers
* [x] Microsoft.AspNetCore.Http.Abstractions
* [x] Microsoft.AspNetCore.Http.Extensions
* [x] Microsoft.AspNetCore.Http.Features
* [x] Microsoft.AspNetCore.Http
* [x] Microsoft.AspNetCore.Metadata
* [x] Microsoft.AspNetCore.Routing.Abstractions
* [x] Microsoft.AspNetCore.Routing
* [x] Microsoft.AspNetCore.WebUtilities
* [x] Microsoft.AspNetCore.Html.Abstractions
* [x] Microsoft.AspNetCore.Identity
* [x] Microsoft.Extensions.Identity.Core
* [x] Microsoft.Extensions.Identity.Stores
* [x] Microsoft.AspNetCore.Connections.Abstractions
* [x] Microsoft.AspNetCore.Server.HttpSys
* [x] Microsoft.AspNetCore.Server.IISIntegration
* [x] Microsoft.AspNetCore.Server.IIS
* [x] Microsoft.AspNetCore.Server.Kestrel.Core
* [x] Microsoft.AspNetCore.Server.Kestrel
* [x] Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets
* [x] Microsoft.AspNetCore.Authentication.Cookies
* [x] Microsoft.AspNetCore.Authentication
* [x] Microsoft.AspNetCore.Authentication.OAuth
* [x] Microsoft.AspNetCore.Authorization
* [x] Microsoft.AspNetCore.Authorization.Policy
* [x] Microsoft.AspNetCore.CookiePolicy
* [x] Microsoft.AspNetCore.Cors
* [x] Microsoft.AspNetCore.Diagnostics.Abstractions
* [x] Microsoft.AspNetCore.Diagnostics
* [x] Microsoft.AspNetCore.Diagnostics.HealthChecks
* [x] Microsoft.AspNetCore.HostFiltering
* [x] Microsoft.AspNetCore.HttpOverrides
* [x] Microsoft.AspNetCore.HttpsPolicy
* [x] Microsoft.AspNetCore.Localization.Routing
* [x] Microsoft.AspNetCore.Localization
* [x] Microsoft.AspNetCore.ResponseCaching.Abstractions
* [x] Microsoft.AspNetCore.ResponseCaching
* [x] Microsoft.AspNetCore.ResponseCompression
* [x] Microsoft.AspNetCore.Rewrite
* [x] Microsoft.AspNetCore.Session
* [x] Microsoft.AspNetCore.StaticFiles
* [x] Microsoft.AspNetCore.WebSockets
* [x] Microsoft.AspNetCore.Razor.Runtime
* [x] Microsoft.AspNetCore.Razor
* [x] Microsoft.AspNetCore.Mvc.Abstractions
* [x] Microsoft.AspNetCore.Mvc.ApiExplorer
* [x] Microsoft.AspNetCore.Mvc.Core
* [x] Microsoft.AspNetCore.Mvc.Cors
* [x] Microsoft.AspNetCore.Mvc.DataAnnotations
* [x] Microsoft.AspNetCore.Mvc.Formatters.Json
* [x] Microsoft.AspNetCore.Mvc.Formatters.Xml
* [x] Microsoft.AspNetCore.Mvc.Localization
* [x] Microsoft.AspNetCore.Mvc.RazorPages
* [x] Microsoft.AspNetCore.Mvc.Razor
* [ ] Microsoft.AspNetCore.Mvc.TagHelpers
* [ ] Microsoft.AspNetCore.Mvc.ViewFeatures
* [x] Microsoft.AspNetCore.Mvc
* [x] Microsoft.AspNetCore.Http.Connections.Common
* [x] Microsoft.AspNetCore.Http.Connections
* [x] Microsoft.AspNetCore.SignalR.Protocols.Json
* [x] Microsoft.AspNetCore.SignalR.Common
* [x] Microsoft.AspNetCore.SignalR.Core
* [x] Microsoft.AspNetCore.SignalR
* [x] Microsoft.AspNetCore.Components.Authorization
* [x] Microsoft.AspNetCore.Components
* [x] Microsoft.AspNetCore.Components.Forms
* [x] Microsoft.AspNetCore.Components.Server
* [x] Microsoft.AspNetCore.Components.Web
* [x] Microsoft.Extensions.FileProviders.Embedded
* [x] Microsoft.Extensions.Caching.SqlServer
* [x] Microsoft.Extensions.Caching.StackExchangeRedis
* [x] Microsoft.Extensions.Configuration.KeyPerFile
* [x] Microsoft.Extensions.Localization.Abstractions
* [x] Microsoft.Extensions.Localization
* [x] Microsoft.Extensions.ObjectPool
* [x] Microsoft.JSInterop
* [x] Microsoft.Extensions.WebEncoders
* [x] Microsoft.Extensions.Diagnostics.HealthChecks.Abstractions
* [x] Microsoft.Extensions.Diagnostics.HealthChecks

## Skipped (potentially no longer commonly used, potential future removal, only used in tests, internal project, etc)

* [x] Microsoft.AspNetCore.Owin
* [x] Microsoft.AspNetCore.Testing
* [x] Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv
* [x] Microsoft.AspNetCore.Authentication.AzureAD.UI
* [x] Microsoft.AspNetCore.Authentication.AzureADB2C.UI
* [ ] Microsoft.AspNetCore.JsonPatch
* [ ] Microsoft.AspNetCore.NodeServices
* [ ] Microsoft.AspNetCore.SpaServices

## Skipped (not for public consumption)
* [x] Microsoft.AspNetCore.DeveloperCertificates.XPlat

### PR DESCRIPTION
**Pull request by pranavkm**
Contributes to https://github.com/dotnet/aspnetcore/issues/5680

### CODE REVIEW

File: src/ObjectPool/ref/Microsoft.Extensions.ObjectPool.netcoreapp.cs [hunk 1]
```diff
@@ -13,6 +13,7 @@ public partial class DefaultObjectPool<T> : Microsoft.Extensions.ObjectPool.Obje
     {
         public DefaultObjectPool(Microsoft.Extensions.ObjectPool.IPooledObjectPolicy<T> policy) { }
         public DefaultObjectPool(Microsoft.Extensions.ObjectPool.IPooledObjectPolicy<T> policy, int maximumRetained) { }
+        [return: System.Diagnostics.CodeAnalysis.NotNullAttribute]
         public override T Get() { throw null; }
         public override void Return(T obj) { }
```
Comment (JamesNK) [inline src/ObjectPool/ref/Microsoft.Extensions.ObjectPool.netcoreapp.cs:18]: Can null be passed to Return? If T is not null everywhere then the notnull generic constraint could be used.

File: src/ObjectPool/ref/Microsoft.Extensions.ObjectPool.netcoreapp.cs [hunk 2]
```diff
@@ -13,6 +13,7 @@ public partial class DefaultObjectPool<T> : Microsoft.Extensions.ObjectPool.Obje
     {
         public DefaultObjectPool(Microsoft.Extensions.ObjectPool.IPooledObjectPolicy<T> policy) { }
         public DefaultObjectPool(Microsoft.Extensions.ObjectPool.IPooledObjectPolicy<T> policy, int maximumRetained) { }
+        [return: System.Diagnostics.CodeAnalysis.NotNullAttribute]
```
Comment (JamesNK) [inline src/ObjectPool/ref/Microsoft.Extensions.ObjectPool.netcoreapp.cs:16]: Should NotNull also be on the base class? Would it ever make sense to return null from ObjectPool?

File: src/ObjectPool/ref/Microsoft.Extensions.ObjectPool.netcoreapp.cs (hunk 1 above)
Comment (pranavkm) [reply src/ObjectPool/ref/Microsoft.Extensions.ObjectPool.netcoreapp.cs:18]: `T` already has a `class` constraint. The compiler requires that both `notnull` and `class` to be the first generic constraint so you're unable to apply both simultaneously. That said, there aren't any null checks here. You could return a null value, but that would likely corrupt the pool

File: src/ObjectPool/ref/Microsoft.Extensions.ObjectPool.netcoreapp.cs (hunk 2 above)
Comment (pranavkm) [reply src/ObjectPool/ref/Microsoft.Extensions.ObjectPool.netcoreapp.cs:16]: It's already on the base class too.

Comment (JamesNK) [reply src/ObjectPool/ref/Microsoft.Extensions.ObjectPool.netcoreapp.cs:16]: I don't see it - https://github.com/dotnet/aspnetcore/blob/42963fbc454e5cc6fedf2389700d73257a97f2a5/src/ObjectPool/ref/Microsoft.Extensions.ObjectPool.netcoreapp.cs#L57-L62

Comment (pranavkm) [reply src/ObjectPool/ref/Microsoft.Extensions.ObjectPool.netcoreapp.cs:16]: Oops, I think I meant to put this on the base type.

File: src/ObjectPool/ref/Microsoft.Extensions.ObjectPool.netcoreapp.cs (hunk 1 above)
Comment (pranavkm) [reply src/ObjectPool/ref/Microsoft.Extensions.ObjectPool.netcoreapp.cs]: As I discovered, `T : class` indicates that `T` is non-null. This removes the need to specify a NotNullAttribute, and we get the right behavior

**Comment by davidfowl** [comment]
Are all the other critical APIs done?

**Comment by pranavkm** [comment]
@JamesNK could I sign off on this?

**Review by JamesNK — APPROVED** [review]

**Review by JamesNK — APPROVED** [review]

**Comment by JamesNK** [comment]
Double approved for extra effectiveness.

**Pull request merged**